networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/RateTracker.java

```diff
@@ -35,27 +39,108 @@
 
   // boundary: if a request comes in and remaining capacity is at least 1, then
   // they can have the objects they request otherwise they get none.
-  public synchronized long popObjectRequests(final long objectCount) {
+  public synchronized Optional<ObjectsRequestResponse> popObjectRequests(final long objectCount) {
     pruneRequests();
+    final UInt64 currentTime = timeProvider.getTimeInSeconds();
     if ((peerRateLimit - requestsWithinWindow) <= 0) {
-      return 0L;
+      return Optional.empty();
     }
 
     requestsWithinWindow += objectCount;
-    requestCount.compute(
-        timeProvider.getTimeInSeconds(),
-        (key, val) -> val == null ? objectCount : val + objectCount);
-    return objectCount;
+    final ObjectsRequestResponse objectsRequestResponse =
+        new ObjectsRequestResponse.ObjectsRequestBuilder(objectCount)
+            .timeSeconds(currentTime)
+            .build();
+    requestCount.put(new ObjectRequestsEntryKey(objectsRequestResponse), objectCount);
+    return Optional.of(objectsRequestResponse);
+  }
+
+  public synchronized void adjustObjectRequests(
+      final ObjectsRequestResponse objectsRequestResponse, final long returnedObjectsCount) {
+    pruneRequests();
+    final ObjectRequestsEntryKey objectRequestsEntryKey =
+        new ObjectRequestsEntryKey(objectsRequestResponse);
+    if (requestCount.containsKey(objectRequestsEntryKey)) {
+      final long initialObjectsCount = requestCount.get(objectRequestsEntryKey);
+      requestCount.put(objectRequestsEntryKey, returnedObjectsCount);
+      requestsWithinWindow = requestsWithinWindow - initialObjectsCount + returnedObjectsCount;
+    }
   }
 
   void pruneRequests() {
     final UInt64 currentTime = timeProvider.getTimeInSeconds();
     if (currentTime.isLessThan(timeoutSeconds)) {
       return;
     }
-    final NavigableMap<UInt64, Long> headMap =
-        requestCount.headMap(currentTime.minus(timeoutSeconds), false);
+    final NavigableMap<ObjectRequestsEntryKey, Long> headMap =
+        requestCount.headMap(
+            new ObjectRequestsEntryKey(currentTime.minus(timeoutSeconds), UUID.randomUUID()),
```

Comment:
And here's the issue taking UUID. We don't know what is minimal uuid.
Let's use AtomicInteger instead.
So add `private final AtomicInteger newRequestId = new AtomicInteger(0)` field to RateTracker, add `requestId(int requrstId)` to builder and use it with `.requestId(newRequestId.getAndIncrement())` when building.
After that init key here with (currentime -...,, 0)

Reply:
Done: Used an `AtomicInteger` to generate request ids instead. Even though when calling `headMap` we only need a sub map based on the time and not the request id. It's better now anyway.
The id generator reset happens when we try to pop requests and we have reached a new time in seconds